### PR TITLE
[bitbucket] enable projects

### PR DESCRIPTION
**Pull request by AlexTugarev**
Resolves https://github.com/gitpod-io/gitpod/issues/5980

This also adds a test of permissions on webhook installations. 

Screen Shot 2021-12-15 at 14 32 29

```release-note
Enables bitbucket.org repositories in Teams & Projects
```

### CODE REVIEW

**Comment by roboquat** [comment]
@AlexTugarev: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **alextugarev** after the PR has been reviewed.

*No associated issue*. Update pull-request body to add a reference to an issue, or get approval with `/approve no-issue`

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)**
- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **alextugarev** after the PR has been reviewed.

Associated issue: *#5980*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)**
- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by AlexTugarev** [comment]
Just found an issue with the cloneURL. 

```
Couldn't parse context 'https://AlexTugarev3@bitbucket.org/AlexTugarev3/tada68.git'.
```

... on it.

**Comment by AlexTugarev** [comment]
@jldec, just fixed the issue with the wrong `cloneUrl`.

**Comment by gtsiolis** [comment]
Looking at this now! 👀

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: 514b037a80e1280716e3a1b70cc1e6a0ba261051</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7251#issuecomment-996743247" title="LGTM">jldec</a>*

Associated issue: *#5980*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)~~ [jldec]
- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [jldec]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [jldec]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Review by gtsiolis — COMMENTED** [review]
Although a bit late but submitting some pending review comments since these could be helpful.

Thanks @AlexTugarev! Feels great to now support Bitbucket repositories as projects. 🔮

File: components/server/src/bitbucket/bitbucket-repository-provider.ts [hunk 1]
```diff
@@ -18,26 +19,82 @@ export class BitbucketRepositoryProvider implements RepositoryProvider {
...
+        })
+
+        for (const branch of response.data.values!) {
+            branches.push({
+                htmlUrl: branch.links?.html?.href!,
+                name: branch.name!,
+                commit: {
+                    sha: branch.target?.hash!,
+                    author: branch.target?.author?.user?.display_name!,
+                    authorAvatarUrl: branch.target?.author?.user?.links?.avatar?.href,
```
[the remark this answers is not in the document]
Comment (AlexTugarev) [reply components/server/src/bitbucket/bitbucket-repository-provider.ts:74]: unfortunately, nope! they don't carry it like here.

File: components/server/ee/src/prebuilds/bitbucket-service.ts [hunk 2]
```diff
@@ -27,7 +27,17 @@ export class BitbucketService extends RepositoryService {
...
+            return false;
+        }
+
+        // only admins may install webhooks on repositories
+        const { owner, repoName } = await this.bitbucketContextParser.parseURL(user, cloneUrl);
+        const api = await this.api.create(user);
+        const response = await api.user.listPermissionsForRepos({
+                q: `repository.full_name="${owner}/${repoName}"`
+            })
+        return !!response.data?.values && response.data.values[0]?.permission === "admin";
```
[the remark this answers is not in the document]
Comment (AlexTugarev) [reply components/server/ee/src/prebuilds/bitbucket-service.ts:40]: Oh, I'll have a look into this. Thanks @gtsiolis!

File: components/server/src/bitbucket/bitbucket-repository-provider.ts [hunk 3]
```diff
@@ -18,26 +19,82 @@ export class BitbucketRepositoryProvider implements RepositoryProvider {
...
+                    authorDate: branch.target?.date!,
+                    commitMessage: branch.target?.message || "missing commit message",
+                }
+            });
+        }
+
+        return branches;
```
[the remark this answers is not in the document]
Comment (gtsiolis) [reply components/server/src/bitbucket/bitbucket-repository-provider.ts:81]: **fyi**: Opened https://github.com/gitpod-io/gitpod/issues/7366 to track this. Cc @AlexTugarev @jldec

File: components/server/ee/src/prebuilds/bitbucket-service.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (gtsiolis) [reply components/server/ee/src/prebuilds/bitbucket-service.ts:40]: **fyi**: Opened https://github.com/gitpod-io/gitpod/issues/7367 to track this. Cc @AlexTugarev @jldec

**Pull request merged**